src/core/lib/surface/completion_queue.c

```diff
@@ -60,26 +60,51 @@ typedef struct {
   void *tag;
 } plucker;
 
+/* Queue that holds the cq_completion_events. This internally uses gpr_mpscq
+ * queue (a lockfree multiproducer single consumer queue). However this queue
+ * supports multiple consumers too. As such, it uses the queue_mu to serialize
+ * consumer access (but no locks for producer access).
+ *
+ * Currently this is only used in completion queues whose completion_type is
+ * GRPC_CQ_NEXT */
+typedef struct grpc_cq_event_queue {
+  /* Mutex to serialize consumers i.e pop() operations */
+  gpr_mu queue_mu;
+
+  gpr_mpscq queue;
+
+  /* A lazy counter indicating the number of items in the queue. This is NOT
+     atomically incremented/decrements along with push/pop operations and hence
+     only eventually consistent */
+  gpr_atm num_queue_items;
+} grpc_cq_event_queue;
+
 /* Completion queue structure */
 struct grpc_completion_queue {
```

Comment:
Can we go ahead and add a vtable for the top half... suggest something like:

```c
struct grpc_cq_vtable {
  size_t size;
  void (*begin_op)(...);
  void (*end_op)(...);
  void (*pluck)(...);
  void (*next)(...);
};
```
and remove data structure bits from grpc_completion_queue (maybe into something like `struct cq_data`)